### ISSUE TITLE
Calling GROUP:Destroy() crashes DCS World

### ISSUE THREAD

**Issue by Loophole62ndFF** [FlightControl-Master/MOOSE#1041]
In DCS World beta Version 2.5.3.22877, if you call GROUP:Destroy() in a "Run Script" Advanced Waypoint Action for a ground unit, DCS World crashes immediately.

E.g. spawn a group consisting of a single APC BTR-80, and in its destination waypoint set it to Run Script:

local vehGrp = GROUP:Find(...)
vehGrp:Destroy()

However, if instead of the above, you do:

local vehGrp = GROUP:Find(...)
local grpUnits = vehGrp:GetUnits()
for i = 1, #grpUnits do
    local grpUnit = vehGrp:GetUnit( i )
    grpUnit:Destroy()
end

then the vehicle is destroyed (but I don't know what happens to the Group object).

(on a side note, the Moose Documentation indicates that Destroy() takes a true/false argument, but the code in Moose.lua does not seem to support that)

**Comment by funkyfranky** [comment]
This is interesting since the Destroy() on group level calls the DCS destroy() on the DCS group object and Destroy() on unit level calls the DCS destroy() event of the DCS unit object. So it might be a DCS bug. However, there are also other events raised in the MOOSE Destroy() function that could be responsible.

As a test, can you try
`vhGrp:Destroy(true)`
or
`vhGrp:Destroy(false)`
and see if DCS still crashes in either of the two cases? This might be helpful to isolate reason for the crash.

Not sure about your side note this the GROUP:Destroy(GenerateEvent) definitely takes and uses a boolean argument - in the docs and in the code.

**Comment by Loophole62ndFF** [comment]
I have tested vhGrp:Destroy(false), vhGrp:Destroy(true) and vhGrp:Destroy().  All three cause a crash.

The DCS log shows the following:

2018-10-24 09:09:13.739 INFO    EDCORE: try to write dump information
2018-10-24 09:09:13.753 INFO    EDCORE: # -------------- 20181024-090912 --------------
2018-10-24 09:09:13.790 INFO    EDCORE: Z:\DCS World OpenBeta\bin\WorldGeneral.dll
2018-10-24 09:09:13.828 INFO    EDCORE: # C0000005 ACCESS_VIOLATION at 84DA13E0 00:00000000
2018-10-24 09:09:13.865 INFO    EDCORE: 00000000 00000000 0000:00000000 
2018-10-24 09:09:13.903 INFO    EDCORE: 84DA13E0 00F8EE20 0000:00000000 Z:\DCS World OpenBeta\bin\WorldGeneral.dll ?onFinishTask_@ControllerMission@AI@@IEAAXH@Z()+20
2018-10-24 09:09:13.940 INFO    EDCORE: 84DA0DF0 00F8EE50 0000:00000000 Z:\DCS World OpenBeta\bin\WorldGeneral.dll ?update@ControllerMission@AI@@UEAAXXZ()+70
2018-10-24 09:09:13.940 INFO    EDCORE: 75D71023 00F8EE80 0000:00000000 Z:\DCS World OpenBeta\bin\Transport.dll ?processMessage@VehicleCommunicator@AI@@MEAAXAEBV?$Ptr@VwMessage@@UDefaultDeleter@ed@@@ed@@@Z()+20F3
2018-10-24 09:09:13.941 INFO    EDCORE: 84D9F8D7 00F8EEE0 0000:00000000 Z:\DCS World OpenBeta\bin\WorldGeneral.dll ?update@Controller@AI@@QEAAXXZ()+107
2018-10-24 09:09:13.941 INFO    EDCORE: 84D44F87 00F8EF10 0000:00000000 Z:\DCS World OpenBeta\bin\WorldGeneral.dll ?Control@wControl@@UEAANXZ()+57
2018-10-24 09:09:13.941 INFO    EDCORE: 75D88AE6 00F8EF40 0000:00000000 Z:\DCS World OpenBeta\bin\Transport.dll ?update@wcColumn@@AEAAXAEAN@Z()+56
2018-10-24 09:09:13.942 INFO    EDCORE: 75D8E021 00F8EF70 0000:00000000 Z:\DCS World OpenBeta\bin\Transport.dll ?setOption@wcColumn@@UEAAXIAEBVParameter@@@Z()+13D1
2018-10-24 09:09:13.942 INFO    EDCORE: B0CF2583 00F8EFF0 0000:00000000 Z:\DCS World OpenBeta\bin\World.dll
2018-10-24 09:09:13.943 INFO    EDCORE: B0CF2A66 00F8F040 0000:00000000 Z:\DCS World OpenBeta\bin\World.dll
2018-10-24 09:09:13.943 INFO    EDCORE: 89725A26 00F8F0B0 0000:00000000 Z:\DCS World OpenBeta\bin\DCS.exe
2018-10-24 09:09:13.943 INFO    EDCORE: 89735FC8 00F8F110 0000:00000000 Z:\DCS World OpenBeta\bin\DCS.exe
2018-10-24 09:09:13.943 INFO    EDCORE: 89710CA0 00F8F140 0000:00000000 Z:\DCS World OpenBeta\bin\DCS.exe
2018-10-24 09:09:13.944 INFO    EDCORE: 89398535 00F8F170 0000:00000000 Z:\DCS World OpenBeta\bin\DCS.exe
2018-10-24 09:09:13.944 INFO    EDCORE: 8939A389 00F8F980 0000:00000000 Z:\DCS World OpenBeta\bin\DCS.exe
2018-10-24 09:09:13.945 INFO    EDCORE: 898BE95F 00F8F9C0 0000:00000000 Z:\DCS World OpenBeta\bin\DCS.exe
2018-10-24 09:09:13.947 INFO    EDCORE: C24C3034 00F8F9F0 0000:00000000 C:\WINDOWS\System32\KERNEL32.DLL BaseThreadInitThunk()+14
2018-10-24 09:09:13.947 INFO    EDCORE: C31E1461 00F8FA40 0000:00000000 C:\WINDOWS\SYSTEM32\ntdll.dll RtlUserThreadStart()+21
2018-10-24 09:09:14.195 INFO    EDCORE: Minidump created.

As an aside, I also noticed in regard to the UNIT:Destroy(); although it works, if you set the event flag to "false", the destroyed unit ends up still counting against the spawn limits (InitLimits) for that template - so can prevent things from spawning when they should.

**Comment by funkyfranky** [comment]
Thanks for testing. If it happens also with the argument set to false, then it strongly suggests that it is a DCS bug. Or (but much more unlikely) a very hidden MOOSE bug. But usually scripting bugs don't cause access violations.

To be 100% sure one would need to destroy the group without using MOOSE, i.e.
```
local myGroup = Group.getByName('tanks')
Group.destroy(myGroup)
```

Regarding the false parameter: it is useful e.g. for re-spawning groups since they do not actually die but are reborn immediately. So in that case you don't want to decrease the number of groups for example in spawn limits but also in various other scenarios. So if you use Destroy(false) you should always know what you are doing. But that's also way the user has to explicitly set the parameter.

**Comment by Loophole62ndFF** [comment]
I tested changing the waypoint action script to just:

Group.destroy(...)

and it still crashed - though I am not certain if that is entirely valid syntax.  I will test again using something closer to your example, and I will also try it with a "normal" (i.e. not SPAWN'd) group and see if it still exhibits the problem.

**Comment by Loophole62ndFF** [comment]
Right, I created an entirely new non-MOOSE mission, with just a single unit (attached).
I verified that the (...) in the waypoint Do Script action references the Group object by doing:
env.info((...).getName(...),true) and getting the group name displayed in the dialog.
I then changed the waypoint Do Script action to be:
Group.destroy(...)
and DCS crashes when it executes the waypoint script.
So it is a DCS bug.  What is the next step - is there somewhere to report this to Eagle Dynamics?

**Comment by Loophole62ndFF** [comment]
I tried calling vehGrp:SetAIOff() before calling vehGrp:Destroy(true), but it still crashes.
Also, my workaround of iterating through the group units, calling UNIT:Destroy() on each one, isn't a viable fix, as it leaves the GROUP entities (in MOOSE and DCS) hanging around and accumulating (I printed out the length of coalition.getGroups() after destroying each set of units; it kept going up)

**Comment by funkyfranky** [comment]
not sure what exactly you want to achieve. maybe/probably there are different ways. for example MOOSE has functions which are executed when a unit reaches a waypoint. could that be an alternative approach?

**Comment by Loophole62ndFF** [comment]
I'm trying to create an endless stream of random vehicles, emerging from a tunnel, driving a loop of road, and disappearing back into another tunnel.  The goal is a "shooting gallery" style mission, where people can compete to achieve a high score in a set period of time, or until some randomly-appearing SAM vehicle manages to shoot them down.  Intended for use at a public display, where we let members of the audience try out our simulator rigs.

**Comment by Delta-99** [comment]
I wouldn't trust attempting to do this via a waypoint action script. Try it in a pure script.

**Comment by thebgpikester** [comment]
Is this issue still prevalent? I never came across it and sounds related to doing it directly in the DCS engine. It sounds like this "DO SCRIPT" is in the Template of a SPAWN, in which case mixing/matching a script called for another possible group in a template is a bit odd but potentially the cause of the reason DCS get's its knickers in a twist because it might be doign an internal check on the existence of the group the command is coming "from".

Also... Find vs FindByName -  principle difference is the DCS Group object is required. Whereas FindByName looks up a name in the database. If it's a template, then its late activated and never spawned, how did the code get it? Just something I was pondering, as I dont use :Find() so I do not know or understand it.

**Comment by Loophole62ndFF** [comment]
I have not revisited this mission in quite some time, but I _have_ had it working in one of the more recent DCS openbeta versions.  I am not sure if the issue was fixed by a DCS change, or because I added a scheduled delay between triggering the script and destroying the group.

The final (working) version uses a "Run Script" advanced action at the final vehicle waypoint, that just does "myMission.RemoveConvoyGroup(...)"

The "..." gets passed as the DCS group id to the script.

The script it calls is this:

function myMission.RemoveConvoyGroup (dcsGroup)
  local grp = GROUP:Find(dcsGroup)
  if grp == nil then
    MESSAGE:New(string.format("Cannot find group %s",tostring(dcsGroup)),30):ToAll()
  else
    myMission.scheduler:Schedule(grp,myMission.DestroyConvoyGroup,{grp},2)
  end
end

function myMission.DestroyConvoyGroup (grp)
  grp:Destroy(False)
end
